tests/test_qc_state_maintenance.py

```
@@ -0,0 +1,185 @@
+    assert state_obj.created_by == "LangQC"
+    assert state_obj.is_preliminary == 1
+
+    state_obj.seq_product.seq_platform.name == "PacBio"
+    sub_product = state_obj.seq_product.product_layout[0].sub_product
+    assert sub_product.properties == json
+    assert sub_product.properties_digest == id
+    assert sub_product.value_attr_one == "Run 1"
+    assert sub_product.value_attr_two == "A"
+
+    id_seq_product = state_obj.id_seq_product
+    hist_obj = (
+        session.execute(
+            select(QcStateHist).filter(QcStateHist.id_seq_product == id_seq_product)
+        )
+        .scalars()
+        .one()
+    )
+    _test_hist_object(state_obj, hist_obj)
+
+    # Current QC state is defined now
+    assert helper.current_qc_state() == state_obj
+    # ... but not for a library
+    assert helper.current_qc_state("library") is None
+
+    args = {
+        "user": users[1],
+        "qc_state": "Passed",
+        "qc_type": "sequencing",
+        "is_preliminary": True,
+        "application": "MyScript",
+    }
+
+    state_obj = helper.assign_qc_state(**args)
+    session.execute(
+        select(QcState).filter(QcState.id_seq_product == id_seq_product)
+    ).scalars().one()  # will error if multiple rows
+
+    assert state_obj.user.username == user2
+    assert state_obj.qc_type.qc_type == "sequencing"
+    assert state_obj.qc_state_dict.state == "Passed"
+    assert state_obj.created_by == "MyScript"
+    assert state_obj.is_preliminary == 1
+    assert state_obj.id_seq_product == id_seq_product
+
+    hist_objs = (
+        session.execute(
+            select(QcStateHist).filter(QcStateHist.id_seq_product == id_seq_product)
+        )
+        .scalars()
+        .all()
+    )
+
+    assert len(hist_objs) == 2
+    _test_hist_object(state_obj, hist_objs[1])
+
+    # Expect no update despite a different user
+    args["user"] = users[0]
+    state_obj_the_same = helper.assign_qc_state(**args)
```

This is a silent failure? No exceptions?

---

At this level we do not care about the users being changed. At the level of the endpoit this will cause an error. Between the two I am not sure how we will implement 'stealing' QC. We'll think about it when somebody actually requests the 'steal' feature.